cmd/capibmadm/cmd/powervs/port/list.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package port
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+
+	powerClient "github.com/IBM-Cloud/power-go-client/clients/instance"
+
+	logf "sigs.k8s.io/cluster-api/cmd/clusterctl/log"
+
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/cmd/capibmadm/clients/iam"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/cmd/capibmadm/clients/powervs"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/cmd/capibmadm/options"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/cmd/capibmadm/printer"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/cmd/capibmadm/utils"
+)
+
+// ListCommand powervs port list command.
+func ListCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "list",
+		Short: "List PowerVS ports",
+		Example: `
+ # List ports in PowerVS Network
+ export IBMCLOUD_API_KEY=<api-key>
+ capibmadm powervs port list --service-instance-id <service-instance-id> --zone <zone> --network <network-name/network-id>`,
+	}
+
+	var network string
+
+	cmd.Flags().StringVar(&network, "network", "", "Network ID or Name(preference will be given to the ID over Name)")
+	_ = cmd.MarkFlagRequired("network")
+
+	cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		return listPorts(cmd.Context(), network)
+	}
+
+	options.AddCommonFlags(cmd)
+	return cmd
+}
+
+func listPorts(ctx context.Context, network string) error {
+	log := logf.Log
+	log.Info("Listing PowerVS ports", "service-instance-id", options.GlobalOptions.ServiceInstanceID, "network", network)
+
+	auth := iam.GetIAMAuth()
+	accountID, err := utils.GetAccountID(ctx, auth)
+	if err != nil {
+		return err
+	}
+	sess, err := powervs.NewPISession(accountID, options.GlobalOptions.PowerVSZone, options.GlobalOptions.Debug)
+	if err != nil {
+		return err
+	}
+
+	networkClient := powerClient.NewIBMPINetworkClient(ctx, sess, options.GlobalOptions.ServiceInstanceID)
+	networks, err := networkClient.GetAll()
+	if err != nil {
+		return fmt.Errorf("failed to get the networks, err: %v", err)
+	}
+
+	var netID string
+
+	for _, net := range networks.Networks {
+		if *net.Name == network || *net.NetworkID == network {
+			netID = *net.NetworkID
+			break
+		}
+	}
+
+	if netID == "" {
+		return fmt.Errorf("not able to find network: \"%s\" by ID or name in the network list", network)
+	}
+
+	portListResp, err := networkClient.GetAllPorts(netID)
```

Comment:
Works with both id and name, hence suggest you to use `network` directly here.

Reply:
Here is a snippet of `GetAllPorts` func, seem it expects `WithNetworkID(id)` @yussufsh am i missing anything here?
```
// Get All Ports on a Network
func (f *IBMPINetworkClient) GetAllPorts(id string) (*models.NetworkPorts, error) {
	params := p_cloud_networks.NewPcloudNetworksPortsGetallParams().
		WithContext(f.ctx).WithTimeout(helpers.PIGetTimeOut).
		WithCloudInstanceID(f.cloudInstanceID).WithNetworkID(id)
```

Reply:
I have tried running the curl command with network name as well. It works for network port object.

Reply:
```
# network_id=ocp-net
# curl -X GET https://syd.power-iaas.cloud.ibm.com/pcloud/v1/cloud-instances/${CLOUD_INSTANCE_ID}/networks/${network_id}/ports  -H 'Authorization: Bearer <token>'   -H 'CRN:<crn>' -H 'Content-Type: application/json'
{"ports":[{"description":"","href":"/pcloud/v1/cloud-instances/<snipped>/networks/2c24ec46-96d2-4dd9-a3f0-f1f96c7e4342/ports/644863e1-2f97-4339-8ec1-3378490a0cf1","ipAddress":"192.168.200.156","macAddress":"fa:19:9e:27:fb:20","portID":"644863e1-2f97-4339-8ec1-3378490a0cf1","pvmInstance":{"href":"/pcloud/v1/cloud-instances/<snipped>/pvm-instances/b687cd71-313d-41df-bfa5-b5f04c5e4575","pvmInstanceID":"b687cd71-313d-41df-bfa5-b5f04c5e4575"},"status":"DOWN"}]}
```

Reply:
@yussufsh this ain't working via code, am I on right track here?
```
prajyotparab@Prajyots-MacBook-Pro capibmadm % ./capibmadm powervs port list --service-instance-id 3229a94c-af54-4212-bf60-6202b6fd0a07 --zone osa21 --network capi-test --debug
Listing PowerVS ports service-instance-id="3229a94c-af54-4212-bf60-6202b6fd0a07" network="capi-test"
GET /pcloud/v1/cloud-instances/3229a94c-af54-4212-bf60-6202b6fd0a07/networks/capi-test/ports HTTP/1.1
Host: osa.power-iaas.cloud.ibm.com
User-Agent: Go-http-client/1.1
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Crn: crn:v1:bluemix:public:power-iaas:osa21:a/c265c8cefda241ca9c107adcbbacaa84:3229a94c-af54-4212-bf60-6202b6fd0a07::
Accept-Encoding: gzip


HTTP/2.0 500 Internal Server Error
Content-Length: 448
Cf-Cache-Status: DYNAMIC
Cf-Ray: 797483d70c1c855b-BOM
Content-Type: application/json
Date: Fri, 10 Feb 2023 11:30:51 GMT
Server: cloudflare
Strict-Transport-Security: max-age=15724800; includeSubDomains

{"description":"unable to get network 9906818c-4ad1-4cd9-b6ca-bba776f2a22d: unable to get network for 9906818c-4ad1-4cd9-b6ca-bba776f2a22d: Resource not found: [GET https://192.168.40.140:9696/v2.0/networks/9906818c-4ad1-4cd9-b6ca-bba776f2a22d], error message: {\"NeutronError\": {\"type\": \"NetworkNotFound\", \"message\": \"Network 9906818c-4ad1-4cd9-b6ca-bba776f2a22d could not be found.\", \"detail\": \"\"}}","error":"internal server error"}

Error: failed to Get all Network Ports for Network capi-test: [GET /pcloud/v1/cloud-instances/{cloud_instance_id}/networks/{network_id}/ports][500] pcloudNetworksPortsGetallInternalServerError  &{Code:0 Description:unable to get network 9906818c-4ad1-4cd9-b6ca-bba776f2a22d: unable to get network for 9906818c-4ad1-4cd9-b6ca-bba776f2a22d: Resource not found: [GET https://192.168.40.140:9696/v2.0/networks/9906818c-4ad1-4cd9-b6ca-bba776f2a22d], error message: {"NeutronError": {"type": "NetworkNotFound", "message": "Network 9906818c-4ad1-4cd9-b6ca-bba776f2a22d could not be found.", "detail": ""}} Error:internal server error Message:}
```

Reply:
As per @yussufsh's comments, the GetAllPorts command works with network name too. Hence removed the GetAllNetworks call ( and passing network argument directly to portList call) and just running the get network with id/name to check if network exists and capturing the error. 

```
 $  ./capibmadm powervs port list --service-instance-id "b9db6009-30fc-4718-98cb-16ba6fd37b82"  --zone mon01 --network rdr-acm-spoke2-411-mon01-pub-net
Listing PowerVS ports service-instance-id="b9db6009-30fc-4718-98cb-16ba6fd37b82" network="rdr-acm-spoke2-411-mon01-pub-net"
DESCRIPTION   EXTERNAL IP     IP ADDRESS       MAC ADDRESS         PORT ID                                STATUS
              169.54.110.91   192.168.164.91   fa:d4:c3:31:24:20   19712105-d5be-4e97-9fc1-d9f49300ddf9   ACTIVE
 ```
 
 ```
 $ ./capibmadm powervs port list --service-instance-id "b9db6009-30fc-4718-98cb-16ba6fd37b82"  --zone mon01 --network Temp      
Listing PowerVS ports service-instance-id="b9db6009-30fc-4718-98cb-16ba6fd37b82" network="Temp"
Error: failed to perform Get Network Operation for Network id Temp with error [GET /pcloud/v1/cloud-instances/{cloud_instance_id}/networks/{network_id}][404] pcloudNetworksGetNotFound  &{Code:0 Description:item not found: unable to get network, subnet, and possibly public network cidr for network Temp on cloud instance 914b05caac05408694e562586a65478b Error:item not found Message:}
 ```

Reply:
Correct. @Prajyot-Parab makes sense?